packages/core/contracts/oracle/implementation/cross-chain/BeaconOracle.sol

```diff
@@ -0,0 +1,150 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+pragma solidity ^0.6.0;
+pragma experimental ABIEncoderV2;
+
+import "../../interfaces/OracleAncillaryInterface.sol";
+import "../../../common/implementation/MultiRole.sol";
+import "../../interfaces/FinderInterface.sol";
+import "./BridgeInterface.sol";
+import "../Constants.sol";
+
+/**
+ * @title Simple implementation of the OracleInterface used to communicate price request data cross-chain between
+ * EVM networks. Can be extended either into a "Source" or "Sink" oracle. The intention is that the "Source" Oracle
+ * is the originator of price resolution data, secured by the DVM, and the "Sink" Oracle can both receive this price
+ * resolution data and request prices cross-chain to the "Source" Oracle. The "Sink" is designed to be deployed on
+ * non-Mainnet networks and the "Source" should be deployed on Mainnet.
+ */
+abstract contract BeaconOracle is OracleAncillaryInterface, MultiRole {
+    enum Roles {
+        Owner, // The owner manages the other roles.
+        Requester, // Can make price requests.
+        Publisher // Can publish price request results.
+    }
+
+    struct Price {
+        bool isAvailable;
+        int256 price;
+    }
+
+    // The two structs below are used in an array and mapping to keep track of prices that have been requested but are
+    // not yet available.
+    struct QueryIndex {
+        bool isValid;
+        uint256 index;
+    }
+
+    // Represents a (identifier, time) point that has been queried.
+    struct QueryPoint {
+        bytes32 identifier;
+        uint256 time;
+        bytes ancillaryData;
+    }
+
+    // Conceptually we want a (time, identifier, ancillaryData) -> price map.
+    mapping(bytes32 => mapping(uint256 => mapping(bytes => Price))) internal verifiedPrices;
+
+    // The mapping and array allow retrieving all the elements in a mapping and finding/deleting elements.
+    // Can we generalize this data structure?
+    mapping(bytes32 => mapping(uint256 => mapping(bytes => QueryIndex))) internal queryIndices;
```

Comment:
Would it be simpler if we did something similar to the Voting contract where we pre-hash these into a bytes32 to make a single mapping `mapping(bytes32 => Price/QueryIndex)`?

Here's the voting contract implementation: https://github.com/UMAprotocol/protocol/blob/eb55ff2941bd5fb587c5db992a6bd8a89672b626/packages/core/contracts/oracle/implementation/Voting.sol#L907-L913.

Reply:
Is that easier? I think we shoudl consider from a performance standpoint whether that's better than a double nested mapping

Reply:
Performance-wise, I would think they're pretty close, but the difference would likely be negligible. In the nested mapping, you basically do:
```
value = storage_slot(hash(hash(hash(some_constant, identifier), time), ancillary_data))
```

In the _encodePriceRequest method, you would be doing:
```
value = storage_slot(hash(some_constant, hash(identifier, time, ancillary_data)))
```

Technically the only difference should be one hash, but you never know which one will end up being more expensive bytecode-wise after compilation.

However, I would argue that since the perf differences are likely negligible, we should go for readability. Whichever one you find to be more readable/understandable SGTM!

Reply:
I ended up going with the hash approach because I actually think its a bit more readable not not have a triple array

Reply:
+1